oslabs-beta/DockerLocal
# Project list use reducer

## Description:
- This changes our projectList from using the useState React Hook to using the useReducer React Hook. This change was implemented because the useReducer hook is preferred over useState when updating state for complex logic or items with nested sub-values. Future work would include refactoring the related code out of the components and into separate functions in projectHelper.ts

## Changes I Made:
- [X] In `AddRepos.tsx`, replace setProjectList with dispatch in props. The toggleIsIncluded function now calls dispatch with an 'addProject' type and a payload of the newProjectList. 
- [X] In `home.tsx`,  replace setProjectList with dispatch function and useReducer. Import useReducer from React. Create projectListReducer function to handle different action types which can update the projectList state. The first useEffect function in this component now calls dispatch with an action type of 'addProject' and res.projectList payload. Pass down dispatch in props to Sidebar.
- [X] In `ProjectPage.tsx`, `, replace setProjectList with dispatch in props. Pass down dispatch in props to ProjectRepoListItem.
- [X] In `ProjectRepoListItem.tsx`, replace setProjectList with dispatch in props. A new project is now created through dispatch with a type of 'addRepo'.
- [X] In `ProjectSidebar.tsx`, replace setProjectList with dispatch in props. The component now calls dispatch with an 'addProject' type and a payload of the updatedProjectList. 
- [X] In `types.tsx`, replace setProjectList with dispatch in all component types where it was being used. Add projectListActions type which defines the actions that can update projectList and is used for dispatch.

## How to Test:
- `npm install`
- `npm start`
- Click sign in with Github.
- Add a new project by using the 'Add Project' button and modal. 
- Once your project is created, add some repositories to it by clicking the 'Add Repositories' button. 
- Check that cloning functionality functions as it did previously. Check off a couple repositories you wish to clone. Make sure that at least one includes a Dockerfile. 
- Check that the 'Compose File' button functions as it did previously. Check off at least one repository which has a Dockerfile and then click the 'Compose File' button. Your docker-compose.yml file should render as shown in the screenshot below.
- Additionally, you can use the React Dev Tools (Components) to check that the state is properly updated as seen in the right side of the screenshot below. 

![testYml](https://user-images.githubusercontent.com/7851187/87492228-f621d980-c5fe-11ea-9c25-dd098920dd0a.png)
